src/properties.ts

```diff
@@ -0,0 +1,219 @@
+   * nice: {
+   *  type: 'boolean',
+   *  label: 'Nice',
+   *  defaultValue: true,
+   * },
+   */
+  BOOLEAN: 'boolean',
+
+  /**
+   * fruit: {
+   *   type: 'options',
+   *   label: 'Fruit',
+   *   defaultValue: 'apple',
+   *   options: {
+   *     Apple: 'apple',
+   *     Banana: 'banana',
+   *     Cherry: 'cherry',
+   *   },
+   * },
+   */
+  OPTIONS: 'options',
+
+  /**
+   *  birthday: {
+   *   type: 'date',
+   *   label: 'Birthday',
+   *   defaultValue: new Date(),
+   *  },
+   */
+  DATE: 'date',
+
+  /**
+   * color: {
+   *   type: 'color',
+   *   defaultValue: '#000000',
+   * },
+   */
+  COLOR: 'color',
+
+  /**
+   * button: {
+   *  type: 'button',
+   *   onClick: () => {
+   *    ... code to modify some variables
+   *  }
+   * },
+   */
+  BUTTON: 'button',
+
+  /**
+   * otherStyles: {
+   *   type: 'object',
+   *   label: 'Styles',
+   *   defaultValue: {
+   *     border: '2px dashed silver',
+   *     borderRadius: 10,
+   *     padding: 10,
+   *   },
+   * },
+   */
+  OBJECT: 'object',
+
+  /**
+   * items: {
+   *   type: 'array',
+   *   label: 'Items',
+   *   defaultValue: ['Laptop', 'Book', 'Whiskey'],
+   * },
+   */
+  ARRAY: 'array',
+
+  /**
+   * images: {
+   *   type: 'files',
+   *   label: 'Happy Picture',
+   *   accept: 'image/*',
+   *   defaultValue: [
+   *     'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAQAAAC1+jfqAAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAAAmJLR0QA/4ePzL8AAAAHdElNRQfiARwMCyEWcOFPAAAAP0lEQVQoz8WQMQoAIAwDL/7/z3GwghSp4KDZyiUpBMCYUgd8rehtH16/l3XewgU2KAzapjXBbNFaPS6lDMlKB6OiDv3iAH1OAAAAJXRFWHRkYXRlOmNyZWF0ZQAyMDE4LTAxLTI4VDEyOjExOjMzLTA3OjAwlAHQBgAAACV0RVh0ZGF0ZTptb2RpZnkAMjAxOC0wMS0yOFQxMjoxMTozMy0wNzowMOVcaLoAAAAASUVORK5CYII=',
+   *   ],
+   * },
+   */
+  FILES: 'files',
+};
+
+export interface StoryPropertyObject {
+  type?: string;
+
+  /**
+   * label to display next to the field editor
+   * by default uses the property name itself
+   */
+
+  label?: string;
+
+  /**
+   * placehlder for empty properties
+   * either undefined defautValue
+   * or user clears the field
+   */
+  placeholder?: string;
+
+  /**
+   * a default value for the property
+   */
+  defaultValue?: any;
+
+  /**
+   * hide the label from the property editor
+   */
+  hideLabel?: boolean;
+  /**
+   * hide the property editor for this property
+   * will only use the defaultValue
+   */
+
+  hidden?: boolean;
+  /**
+   * allows grouping of the properties
+   * in a property editor
+   * for example different editor tabs
+   */
+  groupId?: string;
+
+  // FIELD TYPE SPECIFIC
+  /**
+   * for button type fields, an onClick handler
+   */
+  onClick?: () => any;
+
+  /**
+   * for options type fields
+   */
+
+  display?: 'select' | 'radio' | 'inline-radio' | 'multi-select' | 'check' | 'inline-check';
+
+  /**
+   * for numeric type fields
+   */
+
+  /**
+   * if true, will display a range type slider editor
+   */
+  range?: boolean;
```

Comment:
I think we need to go a little more advanced typescript here, and make individual interfaces for the types.

a `type = BUTTON` cannot have a property `step` for example.

We can ask some typescript exports like @gaetanmaisse or @kroeder for help!

Reply:
sounds good

Reply:
great, lets me also tag them @gaetanmaisse @kroeder

Reply:
I'd concur. Allows us to solve the defaultValue 'any' problem.

Reply:
I think we end up with something like:

```javascript
export interface StoryPropertyObject<T> {
....
  /**
   * a default value for the property
   */
  defaultValue?: T;

...
}
```

and then:

```javascript
export interface NumberPropertyObject extends StoryPropertyObject<number> {
  /**
   * if true, will display a range type slider editor
   */
  range?: boolean;

  /**
   * minimum allowed value for numeric propery
   */
  min?: number;

  /**
   * maximum allowed value for numeric propery
   */
  max?: number;

  /**
   * stepper for numeric editor /i nc/dec value
   */

  step?: number;
}
```

That also unlocks some magic like:

```javascript
interface OnClickFunc {
  (): any;
}

export interface ButtonPropertyObject extends StoryPropertyObject<OnClickFunc> {}
```

Reply:
You were 💯 right, with this new implementation `StoryProperty` is a [Discriminated Unions](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) 💪